moveit_kinematics/ikfast_kinematics_plugin/scripts/auto_create_ikfast_moveit_plugin.sh

```diff
@@ -0,0 +1,212 @@
+EOF
+	test "$QUIET" == "1" && exec 3>&1 1>/dev/null  # save stdout as 3, then redirect to /dev/null
+	docker build -t fixed-openrave $TMP_DIR || exit 1
+	test "$QUIET" == "1" && exec 1>&3  # restore stdout
+	echo "Successfully built docker image."
+	__DOCKER_BUILT=1
+}
+
+function extract_robot_name {
+	local robot_name=$(check_urdf "$1" 2> /dev/null | grep "^robot name is: ")
+	robot_name=${robot_name##robot name is: }
+	test -z "$robot_name" && return
+	if [ -n "$ROBOT_NAME" -a "$ROBOT_NAME" != "$robot_name" ]; then
+		echo "Robot name in urdf ($robot_name) differs from specified name ($ROBOT_NAME)"
+	else
+		ROBOT_NAME="$robot_name"
+	fi
+}
+
+function create_dae_file {
+	echo "Converting urdf to Collada"
+	if ! rosrun collada_urdf urdf_to_collada "$INPUT" "$DAE_FILE" 2> /dev/null ; then
+		# When this failed, run docker
+		build_docker_image
+		echo "Converting urdf to Collada"
+		cp "$INPUT" "$TMP_DIR/robot.urdf"
+		test "$QUIET" == "1" && exec 3>&2 2>/dev/null  # save stderr as 3, then redirect to /dev/null
+		docker run --rm --user $(id -u):$(id -g) -v $TMP_DIR:/input --workdir /input -e HOME=/input \
+				 fixed-openrave:latest rosrun collada_urdf urdf_to_collada robot.urdf robot.dae
+		test "$QUIET" == "1" && exec 2>&3  # restore stderr
+	fi
+}
+
+function create_solver {
+	# create wrapper.xml file, needed by OpenRave
+	cat <<EOF > $TMP_DIR/wrapper.xml
+<robot file="robot.dae">
+  <Manipulator name="$ROBOT_NAME">
+    <base>$BASE_LINK</base>
+    <effector>$EEF_LINK</effector>
+  </Manipulator>
+</robot>
+EOF
+
+	build_docker_image
+
+	# Assemble openrave command
+	cmd="openrave0.9.py --database inversekinematics --robot=/input/wrapper.xml --iktype=$IK_TYPE --iktests=1000"
+	echo "Running $cmd"
+
+	# save stdout as 3, then redirect to /dev/null
+	test "$QUIET" == "1" && exec 3>&1 1>/dev/null
+	# run $cmd in docker as current user, outputting files to $TMP_DIR/.openrave
+	docker run --rm --user $(id -u):$(id -g) \
+		-v $TMP_DIR:/input --workdir /input -e HOME=/input \
+		fixed-openrave:latest $cmd
+
+	# restore stdout
+	test "$QUIET" == "1" && exec 1>&3
+
+	# update INPUT to generated .cpp
+	INPUT=$(ls -1 $TMP_DIR/.openrave/*/*.cpp 2> /dev/null)
+	if [ -n "$INPUT" ] ; then
+		echo "Created $INPUT"
+	else
+		echo "Failed to create ikfast solver"
+		exit 1
+	fi
+}
+
+function create_plugin {
+	echo
+	echo "Running $(dirname $0)/create_ikfast_moveit_plugin.py"
+	$(dirname "$0")/create_ikfast_moveit_plugin.py "$ROBOT_NAME" "$PLANNING_GROUP" "$PKG_NAME" "$BASE_LINK" "$EEF_LINK" "$INPUT"
+}
+
+### main program ###
+
+parse_options $*
+
+# Create a temporary directory to operate in
+TMP_DIR=$(mktemp -d --tmpdir ikfast.XXXXXX)
+DAE_FILE="$TMP_DIR/robot.dae"
+if [ "$KEEP_RESULTS" == "1" ] ; then
+	echo "Storing intermediate results in: $TMP_DIR"
+else
+	# Register cleanup function to be called on EXIT signal
+	trap cleanup EXIT
+fi
+
+# Depending on input file type, select the required processing stages
+while true ; do
```

Comment:
Is it really that much more efficient to have this  as a while loop? Wouldn't it be better to have 3 functions that can call each other?

Reply:
The key idea of the while (and switch) is that you can enter the generation process at **any** stage.
You can't handle this with a fixed call chain.

Reply:
Can't you? What would happen if you split this up into:
```
function process_cpp
{
  ...
}

function process_dae
{
  ...
  process_cpp
}


function process_urdf
{
  ...
  process_dae
}

if [ $expression == urdf ]; then 
  process_urdf
elif [ $expression == dae ]; then 
  process_dae
elif [ $expression == cpp ]; then 
  process_cpp
fi
```

Reply:
That said, I understand that you have already put a great deal of effort into this PR and I don't feel strongly about this change.